rts/lra/lra-service-base/src/main/java/io/narayana/lra/ResponseHolder.java

```diff
@@ -0,0 +1,84 @@
+package io.narayana.lra;
+
+import io.narayana.lra.logging.LRALogger;
+import org.apache.http.Header;
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.methods.HttpRequestBase;
+import org.apache.http.util.EntityUtils;
+
+import javax.ws.rs.core.HttpHeaders;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+
+public class ResponseHolder {
+    private HttpRequestBase httpRequest;
+    private HttpResponse httpResponse;
+    private int status;
+    private String locationHeader;
+    private String responseString;
+
+    public ResponseHolder(HttpRequestBase httpRequest, HttpResponse httpResponse) {
+        this.httpRequest = httpRequest;
+        this.httpResponse = httpResponse;
+        status = httpResponse.getStatusLine().getStatusCode();
+        Header h = httpResponse.getFirstHeader(HttpHeaders.LOCATION);
+        locationHeader = h == null ? null : h.getValue();
+        HttpEntity entity = httpResponse.getEntity();
+        try {
+            responseString = entity != null ? EntityUtils.toString(entity, "UTF-8") : null;
+        } catch (IOException e) {
+            responseString = null;
+        }
+    }
+
+    public URI getRequestURI() {
+        return httpRequest.getURI();
+    }
+
+    public int getStatus() {
+        return status;
+    }
+
+    public String getLocationHeader() {
+        return locationHeader;
+    }
+
+    public String getResponseString() {
+        return responseString;
+    }
+
+    public URI getLocationHeaderAsURI() {
+        try {
+            return locationHeader != null ? new URI(locationHeader) : null;
+        } catch (URISyntaxException e) {
+            if (LRALogger.logger.isInfoEnabled()) {
+                LRALogger.logger.infof("LRARecord.doEnd missing Location header on ACCEPTED response %s failed: %s",
+                        getRequestURI(), e.getMessage());
+            }
+
+            return null;
+        }
+    }
+
+    public boolean hasEntity() {
+        return responseString != null;
+    }
+
+    public String getLastHeader(String headerName) {
+        return httpResponse.getLastHeader(headerName).getValue();
```

Comment:
possible NPE?

Reply:
I don't think so. Only a non spec compliant (or malicious) coordinator would omit the header.